### PR TITLE
PYIC-3440 GA4 page load

### PR DESCRIPTION
**Pull request by Joe-Edwards-GDS**
### What changed

- Added support for GA4 events, following the pattern from DCMAW, with some adjustments for IPV Core
- Patch the express status code setter to track the status code returned
- Associated config in the deployment

See also: https://github.com/alphagov/di-ipv-dca-front/tree/main/src/assets/javascript/analytics

### Issue tracking

- [PYIC-3440](https://govukverify.atlassian.net/browse/PYIC-3440)

### Environment variables or secrets

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

[PYIC-3440]: https://govukverify.atlassian.net/browse/PYIC-3440?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -589,8 +589,16 @@ Resources:
...
+              Value: !If [ IsProduction, "GTM-K4PBJH3", "GTM-KD86CMZ" ]
             - Name: ANALYTICS_DOMAIN
-              Value: !If [ IsProduction, "account.gov.uk", !Sub "${Environment}.account.gov.uk" ]
+              Value: !If
+                - IsProduction
+                - "account.gov.uk"
+                - !If
+                  - IsNewDevelopment
+                  - !Sub "${Environment}.dev.identity.account.gov.uk"
+                  - !Sub "${Environment}.account.gov.uk"
```
Comment (Joe-Edwards-GDS) [inline deploy/template.yaml:601]: Currently this gets set to (something like) `dev-joee.account.gov.uk` in dev, but it should be `dev-joee.dev.identity.account.gov.uk`.

The new analytics mechanism rechecks the preferences even after accepting, so without this change no analytics will work in dev.

This also explains why it's not properly saving cookie preferences.

File: package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
     "build-sass": "rm -rf dist/public/style.css && sass --no-source-map src/assets/scss/application.scss dist/public/stylesheets/application.css --style compressed",
     "copy-locales": "mkdir -p dist && copyfiles -u 1 src/**/*.njk dist/ src/locales/**/** dist/ ",
     "copy-images": "copyfiles -u 3 src/assets/images/** dist/public/images",
-    "minfiy-build-front-end-gov-js": "mkdir -p dist/public/javascripts; uglifyjs node_modules/govuk-frontend/govuk/all.js src/assets/javascript/application.js -o dist/public/javascripts/application.js -c -m ",
+    "minfiy-build-front-end-gov-js": "mkdir -p dist/public/javascripts; uglifyjs node_modules/govuk-frontend/govuk/all.js src/assets/javascript/**/**/*.js src/assets/javascript/**/*.js -o dist/public/javascripts/application.js -c -m ",
```
Comment (Joe-Edwards-GDS) [inline package.json]: This is a bit clunky - for some reason the `**` glob is only expanding one directory level, so this won't work if any deeper directories are added. Suggestions welcome!

File: src/assets/javascript/analytics/cookieBanner.js
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/analytics/cookieBanner.js:1]: All these new files are copied from DCMAW, except where highlighted

File: src/assets/javascript/analytics/cookies.js
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/analytics/cookies.js:1]: Copied from DCMAW

File: src/assets/javascript/analytics/core.js
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/analytics/core.js:1]: Copied from DCMAW

File: src/assets/javascript/analytics/ga4/pageViewTracker.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        page_view: {
+          language: this.getLanguage(),
+          location: this.getLocation(),
+          organisations: '<OT1056>',
+          primary_publishing_organisation: 'government digital service - digital identity',
+          status_code: this.getStatusCode(),
+          title: this.getTitle(),
+          referrer: this.getReferrer(),
+          taxonomy_level1: 'web cri',
+          taxonomy_level2: 'pre cri'
```
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/analytics/ga4/pageViewTracker.js]: This has been slightly modified, to change the taxonomy and fetch the language from the html attribute

File: src/assets/javascript/analytics/ua/init.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    const languageNames = {
+      "en": "english",
+      "cy": "welsh"
+    };
+
+    sendData({
+      event: "langEvent",
+      language: languageNames[languageCode],
+      languagecode: languageCode
+    });
```
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/analytics/ua/init.js:37]: This has changed to match the old UA behaviour in IPV Core

File: src/assets/javascript/app/init.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+window.DI = window.DI || {};
+
+(function (DI) {
+  'use strict'
+
+  function appInit({
+    analyticsCookieDomain,
+    uaContainerId,
+    ga4ContainerId,
```
Comment (Joe-Edwards-GDS) [inline src/assets/javascript/app/init.js]: This has been simplified so it always loads UA and GA4

File: src/lib/locals.js [hunk 6]
```diff
@@ -1,18 +1,28 @@
...
     res.locals.analyticsCookieDomain = GTM_ANALYTICS_COOKIE_DOMAIN;
     res.locals.assetsCdnPath = CDN_PATH;
     res.locals.assetPath = CDN_DOMAIN + "/assets";
+
+    // Patch the status code setter to make it available in locals as well
+    const setStatusCode = res.status;
+    res.status = function (code) {
+      res.locals.statusCode = code;
+      return setStatusCode.call(res, code);
+    };
```
Comment (Joe-Edwards-GDS) [inline src/lib/locals.js:24]: It'd be nicer to read `res.statusCode` at the point of rendering, but I could work out a good way to do it

File: package.json (hunk 2 above)
Comment (wilsond-gds) [reply package.json]: Should we also correct the spelling mistake in 'minfiy' as well?

Comment (wilsond-gds) [reply package.json]: I tried [uglifyjs-folder](https://www.npmjs.com/package/uglifyjs-folder) and it seemed to work: `uglifyjs-folder app/assets/javascripts/ --output app/assets/javascripts/processed/core.js --pattern "**/*.js"`

File: src/views/shared/base.njk [hunk 7]
```diff
@@ -104,5 +104,13 @@
...
     {% block scripts %}{% endblock %}
     <script nonce='{{ cspNonce }}' src="{{ assetsCdnPath }}/public/javascripts/application.js"></script>
-    <script nonce='{{ cspNonce }}'>window.GOVSignIn.appInit("{{ gtmId }}", "{{ analyticsCookieDomain }}", "{{ googleTagManagerPageId }}")</script>
+    <script nonce='{{ cspNonce }}'>
+        window.DI.httpStatusCode = {{ statusCode | d(200) }};
+        window.DI.journeyState = "{{ googleTagManagerPageId }}";
+        window.DI.appInit({
+            analyticsCookieDomain: "{{ analyticsCookieDomain }}",
+            uaContainerId: "{{ uaContainerId }}",
+            ga4ContainerId: "{{ ga4ContainerId }}",
```
Comment (wilsond-gds) [inline src/views/shared/base.njk:113]: Do we need to also pass the `isGa4Enabled` parameter here? [see the DCMAW documentation](https://github.com/alphagov/di-ipv-dca-front/blob/main/src/assets/javascript/docs/analytics.md#:~:text=We%20then%20call,the%20current%20environment.)

Comment (Joe-Edwards-GDS) [reply src/views/shared/base.njk:113]: I had removed it to keep things simple and only use the new UA + GA4 configuration (letting us delete all the old code) - unless you think it's important to have it?

Comment (wilsond-gds) [reply src/views/shared/base.njk:113]: as long as we're sending UA and GA4 events I can’t see a problem with removing it I think

**Review by wilsond-gds — APPROVED** [review]

**Pull request merged**